lib/Serialization/DeserializationErrors.h

```diff
@@ -335,6 +337,74 @@ class XRefNonLoadedModuleError :
   }
 };
 
+/// Project issue affecting modules. Usually a change in the context between
+/// the time a module was built and when it was imported.
+class ModularizationError : public llvm::ErrorInfo<ModularizationError> {
+  friend ErrorInfo;
+  static const char ID;
+  void anchor() override;
+
+public:
+  enum class Kind {
+    Moved,
+    TypeChanged,
+    NotFound
+  };
+
+private:
+  DeclName name;
+  Kind kind;
+  Identifier expectedModuleName;
+  StringRef referencedFromModuleName;
+  Identifier foundModuleName;
+  XRefTracePath path;
+
+public:
+  explicit ModularizationError(DeclName name,
+                               Kind kind,
+                               Identifier expectedModuleName,
+                               StringRef referencedFromModuleName,
+                               Identifier foundModuleName,
+                               XRefTracePath path):
+    name(name), kind(kind), expectedModuleName(expectedModuleName),
+    referencedFromModuleName(referencedFromModuleName),
+    foundModuleName(foundModuleName), path(path) {}
+
+  void diagnose(ASTContext &ctx,
+                DiagnosticBehavior limit = DiagnosticBehavior::Fatal) const;
+
+  void log(raw_ostream &OS) const override {
+    OS << "modularization issue on '" << name << "', reference from '";
```

Comment:
referenced from

Reply:
I left this as is because it reads `reference from X not resolvable`. In theory nobody should see this error as it is replaced by the proper diagnostic.